source/parser/x86_64/type_checker.hpp

```diff
@@ -27,12 +29,46 @@ namespace smeagle::x86_64 {
   inline bool is_ref(st::dataClass dc) { return dc == st::dataReference; }
 
   // Return true if dataClass is reference or pointer
-  inline bool is_indirect(st::dataClass dc) {
-    return dc == st::dataPointer || dc == st::dataReference;
-  }
+  inline bool is_indirect(st::dataClass dc) { return is_pointer(dc) || is_ref(dc); }
 
   // Return true if dataClass is a primitive type
-  inline bool is_primitive(st::dataClass dc) { return dc == st::dataEnum || dc == st::dataScalar; }
+  inline bool is_primitive(st::dataClass dc) { return dc == st::dataScalar; }
   inline bool is_typedef(st::dataClass dc) { return dc == st::dataTypedef; }
+  inline bool is_enum(st::dataClass dc) { return dc == st::dataEnum; }
+
+  // Dereference a pointer or reference
+  inline st::Type *deref(st::Type *t) {
+    if (is_pointer(t->getDataClass())) {
+      return deref(t->getPointerType()->getConstituentType());
+    }
+    if (is_ref(t->getDataClass())) {
+      return deref(t->getRefType()->getConstituentType());
+    }
+    return t;
+  }
+  // Unwrap and remove typedef
+  inline st::Type *remove_typedef(st::Type *t) {
+    if (is_typedef(t->getDataClass())) {
+      t = t->getTypedefType()->getConstituentType();
+    }
+    return t;
+  };
+
+  namespace detail {
+    inline std::pair<st::Type *, int> dedecorate(st::Type *t, int ptr_cnt) {
```

Comment:
Okay so this is going to return the underlying type and a count of times that it was unwrapped? Is there any reason this function is similar to deref? I'm wondering if the use cases are that different, and if it's possible to have one function that serves both / all use cases.

Reply:
Oh, I forgot to comment this. The count is the number of pointer indirections encountered. I believe we will need this for representing multiple indirection when generating facts. I actually didn't see a mention of this in the ABI model doc, so maybe we should ask Matt about it.

>Is there any reason this function is similar to deref?

This function will replace `deref`. I tried to do that in this PR, but it breaks way too many tests. I'll fix that and the tests in the PR for changing `getRegisterClassFromType` that we were working on today.

Reply:
Ah gotcha, that sounds perfect, and I'm good with removing at a later time. 

@mplegendre see above for @hainest question - should we keep track of the number of multiple indirections?